src/ray/util/url.cc

```diff
@@ -16,64 +16,114 @@
+static std::string ScanToken(std::string::const_iterator &c_str, std::string format) {
+  int i = 0;
+  std::string result;
+  format += "%n";
+  if (static_cast<size_t>(sscanf(&*c_str, format.c_str(), &i)) <= 1) {
+    result.insert(result.end(), c_str, c_str + i);
+    c_str += i;
   }
-  std::string scheme = endpoint.substr(scheme_begin, scheme_end - scheme_begin);
-  RAY_CHECK(scheme_end == host_begin || scheme == "tcp");
-  size_t port_end = endpoint.find('/', host_begin);
-  if (port_end >= endpoint.size()) {
-    port_end = endpoint.size();
+  return result;
+}
+
+std::string endpoint_to_url(
+    const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol> &ep,
+    bool include_scheme) {
+  std::string result;
+  switch (ep.protocol().family()) {
+  case AF_INET:
+  case AF_INET6: {
+    boost::asio::ip::tcp protocol = ep.protocol().family() == AF_INET
+                                        ? boost::asio::ip::tcp::v4()
+                                        : boost::asio::ip::tcp::v6();
+    boost::asio::ip::tcp::endpoint e(protocol, 0);
+    RAY_DCHECK(e.size() == ep.size());
+    memcpy(e.data(), ep.data(), ep.size());
+    std::ostringstream ss;
+    ss << e;
+    result = ss.str();
+    if (include_scheme) {
+      result.insert(0, "tcp://");
+    }
+    break;
   }
-  size_t host_end, port_begin;
-  if (endpoint.find('[', host_begin) == host_begin) {
-    // IPv6 with brackets (optional ports)
-    ++host_begin;
-    host_end = endpoint.find(']', host_begin);
-    if (host_end < port_end) {
-      port_begin = endpoint.find(':', host_end + 1);
-      if (port_begin < port_end) {
-        ++port_begin;
-      } else {
-        port_begin = port_end;
-      }
-    } else {
-      host_end = port_end;
-      port_begin = host_end;
+  case AF_UNIX: {
+    size_t path_offset = offsetof(sockaddr_un, sun_path);
+    result.append(reinterpret_cast<const char *>(ep.data()) + path_offset,
+                  ep.size() - path_offset);
+    if (include_scheme) {
+      result.insert(0, "unix://");
     }
-  } else if (std::count(endpoint.begin() + static_cast<ptrdiff_t>(host_begin),
-                        endpoint.begin() + static_cast<ptrdiff_t>(port_end), ':') > 1) {
-    // IPv6 without brackets (no ports)
-    port_begin = port_end;
-    host_end = port_begin;
+    break;
+  }
+  default:
+    RAY_LOG(FATAL) << "unsupported protocol family: " << ep.protocol().family();
+    break;
+  }
+  return result;
+}
+
+boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>
+parse_url_endpoint(const std::string &endpoint, int default_port) {
+  // Syntax reference: https://en.wikipedia.org/wiki/URL#Syntax
+  // Note that we're a bit more flexible, to allow parsing "127.0.0.1" as a URL.
+  boost::asio::generic::stream_protocol::endpoint result;
+  std::string address = endpoint, scheme;
+  if (address.find("unix://") == 0) {
+    scheme = "unix";
+    address.erase(0, scheme.size() + 3);
```

Comment:
To avoid this arithmetic, let's just user `unix://` as the scheme, same below.

Reply:
Pushed